nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -148,57 +156,80 @@ void KeepoutFilter::process(
     return;
   }
 
-  double wx, wy;  // world coordinates
-
-  // Optimization: iterate only in overlapped
-  // (min_i, min_j)..(max_i, max_j) & mask_costmap_ area.
-  //
-  //           mask_costmap_
-  //       *----------------------------*
-  //       |                            |
-  //       |                            |
-  //       |      (2)                   |
-  // *-----+-------*                    |
-  // |     |///////|<- overlapped area  |
-  // |     |///////|   to iterate in    |
-  // |     *-------+--------------------*
-  // |    (1)      |
-  // |             |
-  // *-------------*
-  //  master_grid (min_i, min_j)..(max_i, max_j) window
-  //
-  // ToDo: after costmap rotation will be added, this should be re-worked.
-
-  // Calculating bounds corresponding to bottom-left overlapping (1) corner
-  int mg_min_x, mg_min_y;  // masger_grid indexes of bottom-left (1) corner
-  // mask_costmap_ -> master_grid intexes conversion
-  const double half_cell_size = 0.5 * mask_costmap_->getResolution();
-  wx = mask_costmap_->getOriginX() + half_cell_size;
-  wy = mask_costmap_->getOriginY() + half_cell_size;
-  master_grid.worldToMapNoBounds(wx, wy, mg_min_x, mg_min_y);
-  // Calculation of (1) corner bounds
-  if (mg_min_x >= max_i || mg_min_y >= max_j) {
-    // There is no overlapping. Do nothing.
-    return;
-  }
-  mg_min_x = std::max(min_i, mg_min_x);
-  mg_min_y = std::max(min_j, mg_min_y);
-
-  // Calculating bounds corresponding to top-right window (2) corner
-  int mg_max_x, mg_max_y;  // masger_grid indexes of top-right (2) corner
-  // mask_costmap_ -> master_grid intexes conversion
-  wx = mask_costmap_->getOriginX() +
-    mask_costmap_->getSizeInCellsX() * mask_costmap_->getResolution() + half_cell_size;
-  wy = mask_costmap_->getOriginY() +
-    mask_costmap_->getSizeInCellsY() * mask_costmap_->getResolution() + half_cell_size;
-  master_grid.worldToMapNoBounds(wx, wy, mg_max_x, mg_max_y);
-  // Calculation of (2) corner bounds
-  if (mg_max_x <= min_i || mg_max_y <= min_j) {
-    // There is no overlapping. Do nothing.
-    return;
+  tf2::Transform tf2_transform;
+  int mg_min_x, mg_min_y;  // masger_grid indexes of bottom-left window corner
+  int mg_max_x, mg_max_y;  // masger_grid indexes of top-right window corner
+
+  if (mask_frame_ != global_frame_) {
+    // Filter mask and current layer are in different frames:
+    // prepare frame transformation if mask_frame_ != global_frame_
+    geometry_msgs::msg::TransformStamped transform;
+    try {
+      transform = tf_->lookupTransform(
+        mask_frame_, global_frame_, tf2::TimePointZero,
+        transform_tolerance_);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR(node->get_logger(), "KeepoutFilter: %s", ex.what());
+      return;
+    }
+    tf2::fromMsg(transform.transform, tf2_transform);
+
+    mg_min_x = min_i;
```

Comment:
If you have different frames, don't you still need to consider the overlapping area of the size of the mask (since its variable)? If you had a mask costmap in the `random_world` frame, it could still be of any arbitrary size to know if there's any overlap at all to do any updates on.

Reply:
See 2 comments below, this would be taken care of if we converted in the callback rather than here

Reply:
> you still need to consider the overlapping area

For differently rotated frames overlapping area might take not only rectangle shape, but rather an arbitrary polygon made from 3 to 8 pinnacles (be it triangle, quadrilateral, hexa- or octagon). The picture below shows the intersection options between `(min_i, min_j, max_i, max_j)` costmap window in `<global_frame>` and `mask` in `<mask_frame>`:
![intersections](https://user-images.githubusercontent.com/60094858/97981750-d0098a80-1de3-11eb-95d0-21efe9b28631.jpg)
From this picture it is seen, that overlapping finding task might be not so trivial as it made for costmap<->mask overlapping in the same frame. So I suggest to remain current calculation area to be restricted by `(min_i, min_j, max_i, max_j)` window and leave search of overlapping as optimization task for the future.

Reply:
> For differently rotated frames overlapping area ...

That is a good point. Let me re-review that chunk of code with that in mind